mlir/lib/Analysis/Presburger/Coalesce.cpp

```diff
@@ -0,0 +1,660 @@
+          i--;
+          break;
+        }
+      } else if (info_2.t &&
+                 info_1.cut.empty()) { // adj_eq Case for one equality
+        SmallVector<int64_t, 8> adjEq;
+        complement(info_2.t.getValue(), adjEq); 
+        if (adjEqCaseNoCut(basicSetVector, j, i,
+                           adjEq)) { // adj_eq noCut cas
+          i--;
+          break;
+        } else if (info_2.t && adjEqCase(basicSetVector, j, i, info_2, info_1,
+                                          false)) { // adjEq case
+          i--;
+          break;
+        }
+      }
+    }
+  }
+  for (size_t i = 0; i < basicSetVector.size(); i++) {
+    new_set.addFlatAffineConstraints(basicSetVector[i]);
+  }
+  return new_set;
+}
+
+void addInequalities(FlatAffineConstraints &bs,
+                     const SmallVector<SmallVector<int64_t, 8>, 8> &inequalities) {
+  for (size_t k = 0; k < inequalities.size(); k++) {
+    bs.addInequality(inequalities[k]);
+  }
+}
+
+void addEqualities(FlatAffineConstraints &bs,
+                   const SmallVector<SmallVector<int64_t, 8>, 8> &equalities) {
+  for (size_t k = 0; k < equalities.size(); k++) {
+    bs.addEquality(equalities[k]);
+  }
+}
+
+bool adjIneqPureCase(SmallVector<FlatAffineConstraints, 4> &basicSetVector,
+                     unsigned i, unsigned j, Info &info_a, Info &info_b) {
+  FlatAffineConstraints newSet(basicSetVector[i].getNumIds(), basicSetVector[i].getNumSymbolIds());
+  addInequalities(newSet, info_a.redundant);
+  addInequalities(newSet, info_a.cut);
+  addInequalities(newSet, info_b.redundant);
+  addInequalities(newSet, info_b.cut);
+  addCoalescedBasicSet(basicSetVector, i, j, newSet);
+  return true;
+}
+
+void addCoalescedBasicSet(SmallVectorImpl<FlatAffineConstraints> &basicSetVector,
+                    unsigned i, unsigned j, FlatAffineConstraints &bs) {
+  if (i < j) {
+    basicSetVector.erase(basicSetVector.begin() + j);
+    basicSetVector.erase(basicSetVector.begin() + i);
+  } else {
+    basicSetVector.erase(basicSetVector.begin() + i);
+    basicSetVector.erase(basicSetVector.begin() + j);
+  }
+  basicSetVector.push_back(bs);
+}
+
+bool protrusionCase(SmallVectorImpl<FlatAffineConstraints> &basicSetVector,
+                    Info &info_a, Info &info_b, unsigned i, unsigned j) {
+  FlatAffineConstraints a = basicSetVector[i];
+  FlatAffineConstraints b = basicSetVector[j];
+  SmallVector<SmallVector<int64_t, 8>, 8> inequalities_b, equalities_b;
+  getBasicSetEqualities(b, equalities_b);
+  getBasicSetInequalities(b, inequalities_b);
+  addAsIneq(equalities_b, inequalities_b);
+  SmallVector<SmallVector<int64_t, 8>, 8> wrapped;
+  for (size_t l = 0; l < info_a.cut.size(); l++) {
+    SmallVector<int64_t, 8> t = info_a.cut[l];
+    FlatAffineConstraints bPrime = FlatAffineConstraints(b.getNumDimIds());
+    addInequalities(bPrime, inequalities_b);
+    shift(t, 1);
+    Simplex simp(bPrime);
+    bPrime.addEquality(t);
```

Comment:
You shift `t` and then add it to `bPrime`, but not to the simplex. If simplex is ready you shift `t` back and stop the loop. Why don't you just shift `t` in the else branch?

Reply:
This was a mistake. The shifted t needs to be added to the simplex, but a non shifted t needs too be pushed onto redundant.